### PR TITLE
B vector interpolation from SkyCoords, skeleton of method (pfsspy.Output.get_Bvec)

### PR DESCRIPTION
**Pull request by STBadman**
Here's an initial skeleton for a method  `pfsspy.Output.get_Bvec` which takes as input a SkyCoord instance and returns the vector B field at that coordinate for the given `pfsspy.Output` object. It can be returned as cartesian or spherical (via `pfsspy.coords.cart2sph`) by specifying the keyword argument `coords_type`

As implemented, the argument is asserted to be type SkyCoord, the keyword argument is asserted to either be "cartesian" or "spherical", and the input SkyCoord is cast to spherical representation (as expected by the interpolator). Caching seems to result in the interpolation occurring slower on the first run after the `Output` isntance is created and quicker after that.

To Do :

- Assertion error messages
- Check the coordinate frame of the input SkyCoord is HeliographicCarrington. I suggest not putting a check on the obstime as it's possible your skycoord might be a spacecraft flying over the Sun with an evolving obstime, while you keep the pfss model the same (i.e. fixed obstime).
- I also suggest not forcing the method to fail if the skycoords are outside the source surface or inside the photosphere, since the interpolator returns nan values if this is the case. Perhaps a warning if all returned values are Nan would work
- Verify the method name, and the inputs and keyword names work. The `coords_type` keyword takes the same values as `representation_type` for `SkyCoord` instances but I made the name distinct to avoid this confusion.
- Write tests and verify the field interpolation produces sensible values and the spherical versions are in the correct convention (see docstring). 
- As it stands, the interpolator expects `skycoord.lon,skycoord.lat,skycoord..radius` to all be 1D vectors. I'm not sure if SkyCoord allows different data shapes so I need to check this and potentially do some flattening to make the interpolation step work if these arrays are not 1D.

### CODE REVIEW

**Comment by STBadman** [comment]
Relevant to #201 and #208

File: pfsspy/__init__.py [hunk 1]
```diff
@@ -512,6 +512,55 @@ def _common_b(self):
...
+        assert isinstance(skycoord,coord.SkyCoord)
+
+        # Ensure representation type is spherical for input to interpolator
+        skycoord.representation_type="spherical"
+
+        # Check coord_type is cartesian or spherical
+        assert coords_type in ["cartesian","spherical"]
+        
+        # Check skycoord coordinate system is Carrington Frame
+
```
Comment (dstansby) [inline pfsspy/__init__.py:547]: The thing to do here is transform the input coordinates into the coordinate frame of the pfss model, something like `skycoord = skycoord.transform_to(self.coordinate_frame)`.

Comment (STBadman) [reply pfsspy/__init__.py:547]: Ah of course! Thanks for the note

**Comment by dstansby** [comment]
Thanks, this will be a great feature to have! Comments:

> To Do :
> 
>     * Assertion error messages
It's more standard for these to be common error messages; `ValueError` if it's a problem with a user supplied argument, or `RuntimeError` if something else has gone wrong. (For the argument checking here I think they would all be `ValueError`)


>     * Check the coordinate frame of the input SkyCoord is HeliographicCarrington. I suggest not putting a check on the obstime as it's possible your skycoord might be a spacecraft flying over the Sun with an evolving obstime, while you keep the pfss model the same (i.e. fixed obstime).

See my inline comment; all we need to do is transform the coordiates to the pfsspy solution frame. The obstime business with Carrington coords is going to be fun an cause people headaches (if my experiences are anything to go by!). What people want they will not get by putting in the output of a heliopy trajectory, because SkyCoord are fixed in inertial space. 99% of the time to get the expected result the `obstime` of the corods should be the same as the `obstime` as the pfss map. So, I recommend warning here if the input frame is HelioGraphicCarrington, and any of the `obstimes` in the coordinates are not equal to the `obstime` of the pfss map.

> 
>     * I also suggest not forcing the method to fail if the skycoords are outside the source surface or inside the photosphere, since the interpolator returns nan values if this is the case. Perhaps a warning if all returned values are Nan would work

I think we should copy scipy convention here (https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html#scipy.interpolate.interp1d) and provide `bounds_error` and `fill_value` keyword arguments.

>     * As it stands, the interpolator expects `skycoord.lon,skycoord.lat,skycoord..radius` to all be 1D vectors. I'm not sure if SkyCoord allows different data shapes so I need to check this and potentially do some flattening to make the interpolation step work if these arrays are not 1D.

Currently the tracing method only accepts 1D, so happy to just accept 1D here too for now and this can be fixed to accept arbitrary shape in the future (see https://github.com/dstansby/pfsspy/issues/188)

This should also get a user-facing example in the examples gallery.

**Comment by STBadman** [comment]
Thanks, will work these suggestions in. I'm realizing the spherical transformation is completely wrong here, I need to construct the spherical basis for each SkyCoord vector to compute the transformation.

Re. an example for the example gallery, I think it might be straightforward to show a from scratch prediction of the Br measured during PSP E1!

**Comment by STBadman** [comment]
Updates 

- Assertion errors replaced with ValueErrors and appropriate messages
- Input `SkyCoord` is transformed to `self.coordinate_frame`
- Conversion to Br, Btheta, Bphi transformation is done properly
- Tests added which assert the output shapes of vector B matches the input shape of the `SkyCoord`, the output are magnetic field vectors in nT and the cartesian to spherical transformation is self consistent for the simpole case where the coordinates where the field is evaluated are along the x axis.

ToDo
- Troubleshoot build issues (passes pytest on my machine)

Questions
- Any other tests needed? 
- Not addressed here : User-facing example. Should this be prepared in the same PR or a different one?
- Is the codecov failure an issue? How can I address them?

**Comment by STBadman** [comment]
The build fail is for python 3.8 with upgrades to sunpy and astropy and seems to be a submodule is astropy (erfa) has been moved to it's own module. I'm not sure how to fix this without breaking the other builds.

**Comment by dstansby** [comment]
I should have fixed the warnings/errors in the master branch.

> Questions
> 
>     * Any other tests needed?

It would be good to check the interpolation is working as expected. Since the dipole field has an alalytic expression at 1rs, could you interpolate a point on the solar surface and check that it has the correct value (for the radial field)

> 
>     * Not addressed here : User-facing example. Should this be prepared in the same PR or a different one?
Lets save that for another one to keep this one a reasonable size.
> 
>     * Is the codecov failure an issue? How can I address them?

If you click through to codecov/patch, then click on the "Diff" tab it wil show you the diff, and green/red boxes on the LH side will show whether each line of code is hit in the tests. It looks like it's just the errors and warnings aren't being tested. You can do something like
```python
with pytest.raises(ValueError, match='error message'):
   code_that_raises_error
```
to check these are raised properly.

**Comment by STBadman** [comment]
Ok, I think I addressed everything above!

The analytic test was interesting - it took me a minute to realize the the inner boundary condition is not a dipole, it is just the radial component of a dipole, and the non radial components are enforced by the combination of this bc and the source surface bc. I tested the full vector at the equator, and found the polar component of the field to be 30% stronger than expected if it was a pure dipole field - that's probably a useful physical insight into how PFSS actually works, will have to think about it some more. In any case I reverted to just verifying Br vanishes interpolated at the solar surface at the equator 

Will work on an example in a separate PR.

**Comment by STBadman** [comment]
Played with it a bit, here is PSP's orbt 1 data (scatter) and prediction (solid line) for an ADAPT-GONG map for Nov 6 2018, and weighting the interpolation by 1/r^2, looks like its working!

image

I realized I didn't include the fill_value and out of bounds errors suggested above.

I also found a breaking error if passing in a `SkyCoord` with `observer=None` where the `.transform_to` throws an error trying to transform to the `pfsspy.Output` coordinate frame. Not quite sure how to handle this, maybe can force the SkyCoord observer to match the input observer if it is None.

**Comment by dstansby** [comment]
👋 just a heads up that this is still on my radar to look at - in a couple of weeks or so I'm going to try and spend a chunk of time on pfsspy to get it to version 1.0, which includes this PR. If I haven't done anything in a month feel free to poke me about this!

File: pfsspy/__init__.py [hunk 2]
```diff
@@ -512,6 +513,74 @@ def _common_b(self):
         self._common_b_cache = br, bs, bp, Sbr, Sbs, Sbp
         return self._common_b_cache
 
+    def get_Bvec(self,skycoord,out_type="cartesian") :
+        """
+        Input
+        =====
```
Comment (dstansby) [inline pfsspy/__init__.py:519]: These should be dashes `-` (and similarly below)

Comment (dstansby) [reply pfsspy/__init__.py:519]: The docstrings all need formatting to match the other docstrings, which are in the numpydoc format: https://numpydoc.readthedocs.io/en/latest/format.html - I'm happy to do this if it's easier?

File: pfsspy/__init__.py [hunk 3]
```diff
@@ -512,6 +513,74 @@ def _common_b(self):
...
+        Magnetic field vectors at the requested locations 
+
+        Keyword Arguments
+        =================
+        coords_type : str
+        Either "cartesian" or "spherical" depending on desired
+        output type :
+            Cartesian -> Bx,By,Bz. 
+            Spherical -> Br, B_theta, B_phi 
+            (physics convention : https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/3D_Spherical.svg/240px-3D_Spherical.svg.png)
```
Comment (dstansby) [inline pfsspy/__init__.py:536]: I'm tempted to keep this simple, and just output spherical coordinates, since the solver is all in spherical coordinates. That would mean avoiding having to deal with coordinate conversion conventions etc. for the cartesian case. Thoughts?

File: pfsspy/__init__.py [hunk 4]
```diff
@@ -512,6 +513,74 @@ def _common_b(self):
         self._common_b_cache = br, bs, bp, Sbr, Sbs, Sbp
         return self._common_b_cache
 
+    def get_Bvec(self,skycoord,out_type="cartesian") :
+        """
```
Comment (dstansby) [inline pfsspy/__init__.py:517]: This needs a sentence or two at the top of the docstring to explain what it does.

**Review by dstansby — COMMENTED** [review]
I think this is looking 👍 . Things to fix:
- I think we should remove the option to get the vectors in cartesian
- The docstring needs tidying up.

File: pfsspy/__init__.py (hunk 2 above)
Comment (STBadman) [reply pfsspy/__init__.py:519]: Thanks David, please go ahead and edit the docstrings.

File: pfsspy/__init__.py (hunk 3 above)
Comment (STBadman) [reply pfsspy/__init__.py:536]: I agree this would be simpler. The reason it's there was the specific application that spurred me to write this PR involved tracing the vectors a long a line of sight through the PFSS model, for which cartesian vectors were much more useful. The transformation is pretty straightforward and could be left to the user. That said, as long as it's a keyword agument with spherical as the default, I think it's ok to have the option - what do you think?

Comment (dstansby) [reply pfsspy/__init__.py:536]: I think the default should definitely be spherical. If there's going to be a cartesian option I think that's fine, but there needs to be a full (mathematical) description of the transformation between the two in the pfsspy docs somewhere.

File: pfsspy/__init__.py (hunk 2 above)
Comment (STBadman) [reply pfsspy/__init__.py]: See latest commits - I think I have conformed to the numpy standards and the docs build looks good

File: pfsspy/__init__.py (hunk 3 above)
Comment (STBadman) [reply pfsspy/__init__.py]: see latest commit, I made spherical the default (I'll just note here that the internal interpolator brgi does output cartesian coordinates), and I added details of the transformation in the docstring

File: pfsspy/__init__.py (hunk 4 above)
Comment (STBadman) [reply pfsspy/__init__.py]: Added a short and longer description and it has propagated to the docs build

**Comment by STBadman** [comment]
Ready for review again I think!

**Review by dstansby — APPROVED** [review]
Looks great, thanks a lot!
